### PR TITLE
Rotating log files only by copy

### PR DESCRIPTION
**Pull request by bizley**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  | #19259

### CODE REVIEW

**Comment by samdark** [comment]
He-he. We did it the other way around in https://github.com/yiisoft/log-target-file/issues/38 :)

**Comment by bizley** [comment]
Oh, only renaming? I though there is a problem for Windows with renaming :)

**Comment by samdark** [comment]
Yeah, need to get to it again and reproduce the issue.

File: framework/log/FileTarget.php [hunk 1]
```diff
@@ -163,7 +165,7 @@ protected function rotateFiles()
                     continue;
                 }
                 $newFile = $this->logFile . '.' . ($i + 1);
-                $this->rotateByCopy ? $this->rotateByCopy($rotateFile, $newFile) : $this->rotateByRename($rotateFile, $newFile);
+                $this->rotateByCopy($rotateFile, $newFile);
```
Comment (mikehaertl) [inline framework/log/FileTarget.php:168]: @bizley Did you also test this on Windows? It uses mandatory locking with `flock()` and I'm not sure if a locked file can be  `unlink()`'ed. A hacky solution could be to release the lock in `rotateFiles()` right before the `unlink()` happens. I'm not sure though.

Comment (bizley) [reply framework/log/FileTarget.php:168]: I did not :/ Hoped that it's covered by the pipe. As I read it seems to be the case that Windows is not deleting the file until the last handle is released but I'm not sure if it throws an error or not. Anyway, question is should I switch back the line for rotating back where it was (after the unlocking) or something else needs to be done.

Comment (mikehaertl) [reply framework/log/FileTarget.php:168]: Moving the line back leaves us with the initial problem: Lock is released too early and while we start rotating in this process a second process can try to start log handling in parallel.

Maybe @stu2162583 (or someone else with PHP on Windows) can help out and do a quick test if a locked file can be deleted under windows from the same process?

I don't know of any other solution - except maybe using a mutex component and not use any file locks here at all.

Comment (bizley) [reply framework/log/FileTarget.php:168]: Quick look at the Monolog - the file is locked only for writing, rotating is done before the writing, for the very moment before the unlink they check whether the file is writable and if so they disable the error handler, unlink the file, and enable handler again.

Comment (mikehaertl) [reply framework/log/FileTarget.php:168]: Seems like they don't use any locking at all - but their whole log handling logic might also be quite different from ours. I don't know.

Still think that we should keep some sort of locking.

**Comment by mikehaertl** [comment]
@bizley Looking at the whole implementation of locking in `FileTarget` again makes me wonder what the idea was here. I don't get, how this should work. If you look at `FileMutex` or the `flock()` examples from the [PHP manual](https://www.php.net/manual/en/function.flock.php) the basic locking logic always goes like:

```php
$file = fopen('file.txt', 'w');
if (flock($file, LOCK_EX)) {
    // do something with file
}
fclose($file);
```

But `FileTarget` does not even check the return value of `flock()`. And given that on Linux locks are *advisory* there's actually - no real locking at all ...!!?

Am I missing something?

**Comment by mikehaertl** [comment]
Ah, there's no `LOCK_NB` flag - so the second process that calls `flock()` in parallel will wait at that point until the lock is available. So locking *does* work.

Sorry for the noise.

**Comment by bizley** [comment]
Hmm, yes, you are right, we have some flawed implementation here. Needs to be fixed.

**Comment by bizley** [comment]
> Ah, there's no `LOCK_NB` flag - so the second process that calls `flock()` in parallel will wait at that point until the lock is available. So locking _does_ work.

Yes, but we should check if locking was successful, right?

**Comment by mikehaertl** [comment]
> Yes, but we should check if locking was successful, right?

It will always be successful - the 2nd process wil simply hang at this point until the lock is obtained. It will never return `false.

**Comment by bizley** [comment]
> It will always be successful

Hmm 🤔 php doc does not suggest that but I'm not an expert. Ok, so what is to be done?

**Comment by mikehaertl** [comment]
My assumption what happens on Windows:

 * `flock()` gives you a mandatory lock on the *file handle*
 * that means operations like `fwrite()`, `ftruncate()` that work on a file handle will be useable on a locked file. We know that this is the case otherwhise we could not write the logs under Windows at all.
 * operations like `unlink()`, `rename()` are probably blocked (to be verified) as they don't use the file handle

What we could do: Instead of deleting the file with `unlink()` we could simply truncate it with `ftruncate($fp, 0)`.

And while we're at it: we should call `fflush($fp)` after `fwrite()` (see example in the PHP manual).

**Comment by bizley** [comment]
:) Do you the time to take over this one?

**Comment by mikehaertl** [comment]
I was thinking the same :D. I may just not find time to test. But I'll open another MR with my suggestion.

**Comment by mikehaertl** [comment]
@bizley I should not do these reviews when I don't have enough time: I see that we already use `ftruncate()` to clear the actual log file. `unlink()` is only used on the other rotated files (which don't have a lock). 

Anyway: I'll open an alternative MR so that we can compare and further discuss the solutions.

**Comment by bizley** [comment]
Closing in favor of #19499

**Pull request closed without merging**